Fix remaining issues in beam score calculation

VsonicV opened the pull request:
# What does this PR do?
This PR further fixes the remaining issues in beam score calculation following #27351 .
More specifically:
1) When adding new hypothesis, the `hyp` in `process` does not include the next generated token on which the current beam score is calculated, but the `hyp` in `finalize` includes all the generated tokens so far. This inconsistency is resolved by changing the `add` function of `BeamHypotheses`. Now we directly pass the current length of the generated tokens to `add`.
2) When calculating best possible beam score in `is_done` function of `BeamHypotheses`, we are directly using `max_length` without deducting `decoder_prompt_len`. It is fixed now.
3) Updated the testing expectation accordingly.


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?


## Who can review?
@gante

VsonicV committed:
Fix issues in add and is_done for BeamHypotheses

VsonicV committed:
make newly added arguments optional for better compatibility

gante commented on src/transformers/generation/beam_search.py:
```diff
@@ -279,15 +279,14 @@ def process(
-                    if decoder_prompt_len == input_ids.shape[-1]:
-                        continue
+                    # add up to the length which the next_scores is calculated on
+                    generated_len = input_ids[batch_beam_idx].shape[-1] + 1 - decoder_prompt_len
```
If I'm not mistaken, this is the same as `cur_len` (L228). I'd suggest renaming `cur_len` into `generated_len`, which is more representative of the variable contents!

VsonicV replied:
Yes, the `input_ids[batch_beam_idx].shape[-1]` should be same for each `batch_beam_idx`, so we can simply use `cur_len` here.

gante commented on src/transformers/generation/beam_search.py:
```diff
@@ -956,12 +954,16 @@ def add(
         Add a new hypothesis to the list.
         """
-        score = sum_logprobs / ((hyp.shape[-1] - decoder_prompt_len) ** self.length_penalty)
+        if generated_len is not None:
```
I'd add a note that the `else` case here exists for retrocompatibility reasons :)

gante reviewed (approved):
Good catch, thanks for fixing!

BTW, run `RUN_SLOW=1 py.test tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::ViT2GPT2ModelIntegrationTest::test_inference_coco_en` -- this test may need an update in its value

VsonicV committed:
Directly use cur_len as generated_len, add note for retrocompatibility

VsonicV committed:
update test expectation

VsonicV commented:
> Good catch, thanks for fixing!
> 
> BTW, run `RUN_SLOW=1 py.test tests/models/vision_encoder_decoder/test_modeling_vision_encoder_decoder.py::ViT2GPT2ModelIntegrationTest::test_inference_coco_en` -- this test may need an update in its value

@gante I have updated the expectation value for this test. I have also incorporated all your suggestions. Ready to go!

VsonicV committed:
make cur_len represents the length of the entire sequence including the decoder prompt

VsonicV commented:
@gante I have also updated the usage of `cur_len` in this Pytorch version following your suggestions in #27814 , now it represents the length of the entire sequence including the decoder prompt, which is consistent with the remaining codebase.

gante commented on tests/generation/test_framework_agnostic.py:
```diff
@@ -634,7 +634,7 @@ def test_eos_token_id_int_and_list_beam_search(self):
             "num_beams": 3,
         }
         if is_pt:
```
nit: we can actually remove this if/else, if the result is back into being the same :P

VsonicV replied:
redundant if/else removed

gante reviewed (approved):
Perfect, thanks for iterating 💛 

Note for @ArthurZucker: there may be slow CI failures due to this change. Although I suspect there won't, since the correction is small. In any case, I'll keep an eye on the CI after this gets merged.

VsonicV committed:
remove redundant if/else in testing

VsonicV commented:
All suggested changes are incorporated. Ready to go! @gante @ArthurZucker

ArthurZucker reviewed (approved):
thanks for this! 🤗

VsonicV commented on src/transformers/generation/beam_search.py:
```diff
@@ -956,12 +948,17 @@ def add(
-        score = sum_logprobs / ((hyp.shape[-1] - decoder_prompt_len) ** self.length_penalty)
+        if generated_len is not None:
+            score = sum_logprobs / (generated_len**self.length_penalty)
+        # This 'else' case exists for retrocompatibility
```
oops, PR already merged, maybe let's stay with it for now?

ArthurZucker replied:
of course no worries
